x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -151,6 +151,25 @@ private <T> void preAnalyze(LogicalPlan parsed, BiFunction<IndexResolution, Enri
         PreAnalyzer.PreAnalysis preAnalysis = preAnalyzer.preAnalyze(parsed);
         Set<String> policyNames = new HashSet<>(preAnalysis.policyNames);
         EnrichResolution resolution = new EnrichResolution(ConcurrentCollections.newConcurrentSet(), enrichPolicyResolver.allPolicyNames());
+        var fieldNames = fieldNames(parsed);
+
+        if (fieldNames == IndexResolver.ALL_FIELDS) {
+            // we don't need the enrich policies match_fields names BEFORE resolving the indices, because all fields are requested anyway
+            preAnalyzeConcurrently(parsed, action, policyNames, resolution, fieldNames, listener);
+        } else {
+            // first we need the match_fields names from enrich policies and THEN, with an updated list of fields, we call field_caps API
+            preAnalyzeSequentially(parsed, action, policyNames, resolution, fieldNames, listener);
```

Comment:
Resolving enrich policies should be trivial since it occurs on the local node with a minimal computation. I think it's better to keep only the sequential version. I think we made it concurrent because it was more convenience rather than necessity.

Reply:
Thanks @dnhatn. I've made it sequential.